maths/fibonacci_sequence_recursion_with_memoization.py

```diff
@@ -0,0 +1,36 @@
+# Fibonacci Sequence Using Recursion and Memoization
+
+# Memoization Concept: https://en.wikipedia.org/wiki/Memoization
+# Fibonacci Concept : https://pt.wikipedia.org/wiki/Sequ%C3%AAncia_de_Fibonacci
+
+cache = dict()
+
+
+def fibonacci(n):
+    """
+    Return number of index n in sequence fibonacci
+    >>> [fibonacci(i) for i in range(12)]
+    [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]
+    """
+    aux = n
+
+    if aux in cache:
+        return cache[aux]
```

Comment:
Please check whether the input is of type `int`, if not then raise `TypeError` and for negative numbers raise `ValueError`.